File: interfaces/python/tests/unittests/warm_start_tests.py
# Test osqp python module
import osqp
# import osqppurepy as osqp
import numpy as np
import scipy.sparse as spspa
import scipy as sp

# Unit Test
import unittest


class warm_start_tests(unittest.TestCase):

    def setUp(self):
        """
        Setup default options
        """
        self.opts = {'verbose': False,
                     'adaptive_rho': False,
                     'eps_abs': 1e-08,
                     'eps_rel': 1e-08,
                     'polish': False,
                     'check_termination': 1}

    def test_warm_start(self):

        # Big problem
        sp.random.seed(2)
        self.n = 100
        self.m = 200
        self.A = spspa.random(self.m, self.n, density=0.9).tocsc()
        self.l = -sp.rand(self.m) * 2.
        self.u = sp.rand(self.m) * 2.

        P = spspa.random(self.n, self.n, density=0.9)
        self.P = P.dot(P.T).tocsc()
        self.q = sp.randn(self.n)

        # Setup solver
        self.model = osqp.OSQP()
        self.model.setup(P=self.P, q=self.q, A=self.A, l=self.l, u=self.u,
                         **self.opts)

        # Solve problem with OSQP
        res = self.model.solve()

        # Store optimal values
        x_opt = res.x
        y_opt = res.y
        tot_iter = res.info.iter

        # Warm start with zeros and check if number of iterations is the same
        self.model.warm_start(x=np.zeros(self.n), y=np.zeros(self.m))
        res = self.model.solve()
        self.assertEqual(res.info.iter, tot_iter)

        # Warm start with optimal values and check that number of iter < 10
        self.model.warm_start(x=x_opt, y=y_opt)
        res = self.model.solve()
        self.assertLess(res.info.iter, 10)

File: interfaces/matlab/unittests/warm_start_tests.m
classdef warm_start_tests < matlab.unittest.TestCase
    %WARM_START_TESTS Warm Start problems solution

    properties
        P
        q
        A
        u
        l
        m,
        n
        solver
        options
        tol
    end

    methods(TestMethodSetup)
        function setup_problem(testCase)

            % Set options
            testCase.options = struct;
            testCase.options.verbose = 0;
            testCase.options.rho = 0.01;
            testCase.options.eps_abs = 1e-04;
            testCase.options.eps_rel = 1e-04;
            testCase.options.adaptive_rho = 0;
            testCase.options.check_termination = 1;

            % Setup tolerance
            testCase.tol = 1e-04;

        end
    end

    methods (Test)
        function test_warm_start(testCase)

            % big example
            rng(4)
            testCase.n = 100;
            testCase.m = 200;
            Pt = sprandn(testCase.n, testCase.n, 0.6);
            testCase.P = Pt' * Pt;
            testCase.q = randn(testCase.n, 1);
            testCase.A = sprandn(testCase.m, testCase.n, 0.8);
            testCase.u = 2*rand(testCase.m, 1);
            testCase.l = -2*rand(testCase.m, 1);

            % Setup solver
            testCase.solver = osqp;
            testCase.solver.setup(testCase.P, testCase.q, ...
                testCase.A, testCase.l, testCase.u, testCase.options);

            % Solve with OSQP
            results = testCase.solver.solve();

            % Store optimal values
            x_opt = results.x;
            y_opt = results.y;
            tot_iter = results.info.iter;

            % Warm start with zeros and check if the number of iterations is the same
            testCase.solver.warm_start('x', zeros(testCase.n, 1), 'y', zeros(testCase.m, 1));
            results = testCase.solver.solve();
            testCase.verifyEqual(results.info.iter, tot_iter, 'AbsTol', testCase.tol)

            % Warm start with optimal values and check that number of iterations is < 10
            testCase.solver.warm_start('x', x_opt, 'y', y_opt);
            results = testCase.solver.solve();
            testCase.verifyThat(results.info.iter, matlab.unittest.constraints.IsLessThan(10));

        end

    end

end
